### PR TITLE
chore: Add HTML/CSS/JS for button screenshot tests

### LINKED ISSUE #1995: Create simple Button screenshot HTML/CSS
- [ ] Create a new simpler webpack.config.js in test/screenshot
- [ ] Screenshot HTML/CSS for button
- [ ] Ability to "run" webpack locally and view the HTML/CSS for button screenshot test

Note that there is no actual "test" yet. This is just setting up a sever

### PR DESCRIPTION
**Pull request by acdvorak**
Resolves #1995

Includes test pages for baseline styles and all mixins.

### How to run it

1. Run:
    ```bash
    npm start
    ```
2. Open http://localhost:8080/ in a browser.

### Example test page

localhost_8080_mdc-button_classes_baseline html

### CODE REVIEW

**Comment by acdvorak** [comment]
@aprigogin suggested that I created small, separate pages because it reduces visual noise when an element's height/padding/margin changes.

If you have one giant HTML file with every variant of a component, and you make a PR that adds 2px to the first variant's `padding-top`, the diff is gonna be huge and noisy because all of the other elements below it will shift down by 2px. This makes it hard to tell what _actually_ changed.

IIRC he encountered this problem frequently on Google News, which is why he recommended to me (rather strongly, I might add) that I should avoid the problem by creating small, isolated pages.

As you said, you end up with a lot more HTML files, screenshot images, and JSON changes, which isn't ideal, but he seemed to think it was worth the trade-off.

**Comment by patrickrodee** [comment]
@acdvorak I see, that makes sense. That way, each test compares a parameter of customization as opposed to having all that jammed in one massive page.

File: scripts/build/static-server.js [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+      // eslint-disable-next-line no-unused-vars
+      filter: (filename, index, files, dir) => {
+        return this.shouldShowFile_(dir, filename, fileExtensions);
+      },
+      stylesheet: stylesheetAbsolutePath,
+      icons: true,
+    };
+
+    const absolutePath = this.pathResolver_.getAbsolutePath(path);
+    app.use(
```
[the remark this answers is not in the document]
Comment (patrickrodee) [reply scripts/build/static-server.js]: That's part of [Express](https://expressjs.com/en/4x/api.html#app.use).

File: test/screenshot/mdc-button/button.test.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as dom from '../../../demos/dom.js';
+import * as pony from '../../../demos/ponyfill.js';
+
+// Prevent default navigation behavior on all <a href="#"> links
```
[the remark this answers is not in the document]
Comment (acdvorak) [reply test/screenshot/mdc-button/button.test.js]: Definitely! I originally did that, but ended up moving it into button.test.js to avoid adding yet another file. I'll add it back.

File: test/screenshot/fixture.test.scss [hunk 3]
```diff
@@ -0,0 +1,35 @@
+//
```
[the remark this answers is not in the document]
Comment (acdvorak) [reply test/screenshot/fixture.test.scss]: At previous companies, I saw the term "fixture" used in screenshot tests to refer to the minimal UI that _hosts_ (wraps or contains) the components being tested.

The word _might_ be borrowed from [JUnit](https://github.com/junit-team/junit4/wiki/test-fixtures)? (Many of my coworkers were Java developers.)

> The purpose of a test fixture is to ensure that there is a well known and fixed environment in which tests are run so that results are repeatable.

I'm open to suggestions for a better name 😀

File: test/screenshot/mdc-button/baseline/text-only.test.html [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+  </head>
+
+  <body class="mdc-typography">
+    <main class="test-main">
+      <p class="test-row">
+        <span class="test-cell">
+          <button class="mdc-button">Button</button>
+        </span>
+        <span class="test-cell">
+          <button class="mdc-button mdc-button--dense">Dense</button>
```
[the remark this answers is not in the document]
Comment (acdvorak) [reply test/screenshot/mdc-button/baseline/text-only.test.html]: What exactly is the definition of "baseline", then? I've seen it used somewhat inconsistently.

Are raised, unelevated, and stroked buttons considered "baseline"? If so, what's the difference between those and "dense"? They're all variants of a button that we expose with a `--modifier` class.

And what about `with-icon.html`? It has dense buttons too. Are you thinking we should have 4 files?

```
text-only.test.html
text-only-dense.test.html
with-icon.test.html
with-icon-dense.test.html
```

File: scripts/build/static-server.js [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+   * @param {number} port
+   * @private
+   */
+  logRunning_(port) {
+    const message = `Local development server running on http://localhost:${port}/`;
+    const ch = '=';
+    const border = ch.repeat(message.length + 8);
+    const spacer = ' '.repeat(message.length);
+
+    console.log(colors.green.bold(`
```
[the remark this answers is not in the document]
Comment (acdvorak) [reply scripts/build/static-server.js]: > documentation about the format of logs

Not sure what you mean by this - can you give me an example?

This is the only log output I have so far, so there's not much to document.

Do you want me to copy/paste the output of the function into the description?

File: scripts/build/static-server.js [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+  /**
+   * Determines whether the given filesystem path (parentPath + filename) should be listed in the directory index page.
+   * @param {string} parentPath
+   * @param {string} filename
+   * @param {!Array<string>} fileExtensions
+   * @return {boolean}
+   * @private
+   */
+  shouldShowFile_(parentPath, filename, fileExtensions) {
+    // No whitelisted file extensions were specified, so show all files
```
[the remark this answers is not in the document]
Comment (acdvorak) [reply scripts/build/static-server.js]: Maybe "whitelist" is the wrong word... I suppose it's kinda misleading.

Ignoring the word choice, the idea itself is quite common.

For example, consider the `ls` shell command: Calling it with no arguments prints _all_ files, whereas passing it a glob pattern only prints files that match the pattern.

```bash
$ ls
dir1
dir2
file.txt
page.html
...

$ ls *.txt
doc.txt
```

In other words, if you don't give `ls` a "filter", it prints _everything_.

Same idea here!

Would it be less confusing if I removed the word "whitelist" altogether?
Is there a better word I could use instead?

File: scripts/build/static-server.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (acdvorak) [reply scripts/build/static-server.js]: Yeah, blame Express. It's classic Node/jQuery/Unix minimalism. I'm not a fan either. 😠 

It's like `atoi()`. What the hell is _that_?? (Googles it) Ohh.... "ASCII to Integer". [_Obviously_](http://www.reactiongifs.com/r/obvs.gif).

File: scripts/build/static-server.js [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+    this.fsLib_ = fsLib;
+    this.serveIndexLib_ = serveIndexLib;
+  }
+
+  /**
+   * Starts a static file server asynchronously and returns immediately.
+   * @param {string} path Relative or absolute path to a filesystem directory to serve
+   * @param {number} port
+   * @param {!Array<string>=} fileExtensions Whitelisted file extensions to show in the directory index
+   * @param {string} stylesheetAbsolutePath Absolute path to a CSS file containing styles for the directory index
```
[the remark this answers is not in the document]
Comment (acdvorak) [reply scripts/build/static-server.js]: Yes, they all have the same CSS file.

It's not about making it customizable, it's about keeping all (or most) of the file paths in one place (`webpack.config.js`) so that our file structure is easier to understand and maintain.

File: scripts/build/static-server.js [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+    this.expressLib_ = expressLib;
+    this.fsLib_ = fsLib;
+    this.serveIndexLib_ = serveIndexLib;
+  }
+
+  /**
+   * Starts a static file server asynchronously and returns immediately.
+   * @param {string} path Relative or absolute path to a filesystem directory to serve
+   * @param {number} port
+   * @param {!Array<string>=} fileExtensions Whitelisted file extensions to show in the directory index
```
[the remark this answers is not in the document]
Comment (acdvorak) [reply scripts/build/static-server.js]: `static-server.js` is located in `scripts/build/`, which is a generic directory.

There's no reason to make this class screenshot test-specific. A static server is inherently generic and reusable.

What's more, it's probably a good idea to keep as much of our _configuration_ as possible in one place (`webpack.config.js`).

For example, we might decide that we want the directory listing to show `.css` and `.js` files, too. Idk about you, but the first place I would look for that is `webpack.config.js`.

File: scripts/build/static-server.js [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+    path,
+    port,
+    directoryIndex: {
+      fileExtensions = [],
+      stylesheetAbsolutePath,
+    } = {},
+  }) {
+    const app = this.expressLib_();
+    const indexOpts = {
+      // eslint-disable-next-line no-unused-vars
```
[the remark this answers is not in the document]
Comment (acdvorak) [reply scripts/build/static-server.js]: Removed. Good catch 😀 

Originally, I wasn't using the `dir` param, but I still wanted to document the entire argument list for future readers and maintainers, which is why I needed the `eslint-disable` comment.

But now that I'm actually using the `dir` param, the comment isn't not necessary anymore.

File: package.json [hunk 10]
```diff
@@ -5,21 +5,23 @@
   "scripts": {
     "build": "npm run clean && mkdirp build && webpack --progress --colors && webpack --config=demos/webpack.config.js --progress --colors",
     "build:min": "mkdirp build && cross-env MDC_ENV=production webpack -p --progress --colors",
+    "build:screenshot": "npm run clean && webpack --config=test/screenshot/webpack.config.js --progress --colors",
```
[the remark this answers is not in the document]
Comment (acdvorak) [reply package.json]: I'll remove `build:screenshot` until we have a real-world need for it.
I was trying to think too far ahead 🙂 

`dev:screenshot` needs `MDC_STATIC_SERVER=true webpack` because it's calling a single shared config file that needs to support multiple "feature flags" (e.g., "start up the dev server" vs. "upload screenshot assets to GCS").

`build:screenshot` didn't need the flag because it wasn't starting up a local dev server - it was only compiling and outputting the static assets for it, which _always_ happens when you run `webpack.config.js`.

File: test/screenshot/mdc-button/custom/all.test.html [hunk 11]
```diff
@@ -0,0 +1,78 @@
+<!DOCTYPE html>
```
[the remark this answers is not in the document]
Comment (acdvorak) [reply test/screenshot/mdc-button/custom/all.test.html]: I disagree. It's the only page that shows this icon positioning bug:

image

File: test/screenshot/mdc-button/button.test.scss [hunk 12]
```diff
@@ -0,0 +1,64 @@
+//
```
[the remark this answers is not in the document]
Comment (acdvorak) [reply test/screenshot/mdc-button/button.test.scss]: So you want:

```
corner-radius.test.scss
fill-color.test.scss
...etc
```

Is that right?

File: test/screenshot/webpack.config.js [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+
+const pathResolver = new PathResolver();
+const globber = new Globber({pathResolver});
+const pluginFactory = new PluginFactory({globber});
+const copyrightBannerPlugin = pluginFactory.createCopyrightBannerPlugin();
+const cssBundleFactory = new CssBundleFactory({env, pathResolver, globber, pluginFactory});
+const jsBundleFactory = new JsBundleFactory({env, pathResolver, globber, pluginFactory});
+const staticServer = new StaticServer({pathResolver});
+
+const MAIN_OUTPUT = {
```
[the remark this answers is not in the document]
Comment (acdvorak) [reply test/screenshot/webpack.config.js]: `const` !== CONSTANT. From the [Google JavaScript Style Guide](https://google.github.io/styleguide/jsguide.html#naming-constant-names):

> Constant names use CONSTANT_CASE: all uppercase letters, with words separated by underscores.
>
> Every constant is a `@const` static property or a module-local `const` declaration, but not all `@const` static properties and module-local `const`s are constants. Before choosing constant case, consider whether the field really feels like a _deeply immutable_ constant. For example, if any of that instance's observable state can change, it is almost certainly not a constant. Merely intending to never mutate the object is generally not enough.

File: test/screenshot/webpack.config.js [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+  fsDirAbsolutePath: pathResolver.getAbsolutePath('/test/screenshot/out/main'),
+};
+
+const TEST_OUTPUT = {
+  fsDirAbsolutePath: pathResolver.getAbsolutePath('/test/screenshot/out/test'),
+};
+
+module.exports = [
+  mainJsCombined(),
+  mainCssALaCarte(),
```
[the remark this answers is not in the document]
Comment (acdvorak) [reply test/screenshot/webpack.config.js]: Speed. Recompiling our combined JS takes about 1 second, whereas compiling our combined Sass takes 15 seconds.

If Sass compilation was faster, I probably would have used combined for CSS too.

File: test/screenshot/webpack.config.js [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+}
+
+function testJs() {
+  return jsBundleFactory.createCustomJs({
+    bundleName: 'screenshot-test-js',
+    chunkGlobConfig: {
+      inputDirectory: '/test/screenshot',
+      filePathPattern: '**/*.test.js',
+    },
+    output: Object.assign({}, TEST_OUTPUT, {
```
[the remark this answers is not in the document]
Comment (acdvorak) [reply test/screenshot/webpack.config.js]: Woops! Copy/paste error. Fixed!

File: test/screenshot/mdc-button/custom/all.test.html (hunk 11 above)
[the remark this answers is not in the document]
Comment (kfranqueiro) [reply test/screenshot/mdc-button/custom/all.test.html]: I might be inclined to agree with Lynn because some of the use cases this is testing seem invalid, which might also explain the "bug"...

For instance, you're applying the stroke mixins to buttons that don't include the `--stroked` class, which means that they have no `border-style` so there is no border and it therefore does not take up space, however the `stroke-width` mixin is still going to reduce padding under the assumption that there is a border applied.

TL;DR the button mixins only make sense when used in conjunction with the relevant modifier classes.

File: scripts/build/static-server.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (patrickrodee) [reply scripts/build/static-server.js]: Could change it to `app.get('/', ...)` to be a *tiny* bit more clear.

File: test/screenshot/mdc-button/custom/all.test.html (hunk 11 above)
[the remark this answers is not in the document]
Comment (acdvorak) [reply test/screenshot/mdc-button/custom/all.test.html]: Good point, Ken. I'll remove this file.

File: scripts/build/static-server.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (acdvorak) [reply scripts/build/static-server.js]: I tried `app.get('/', ...)` but it didn't work - I got 404's for all subdirectories.

I think `use` is lower level than `get` (in fact `get` calls `use` under the hood).

The static middlewares (`this.expressLib_.static(absolutePath)` and `this.serveIndexLib_(absolutePath, indexOpts)`) probably need lower level access... or something...?

I added a comment above this line explaining what it's for.

File: test/screenshot/mdc-button/button.test.scss [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+$custom-button-color: $material-color-red-300;
+$custom-button-contrast-color: mdc-theme-accessible-ink-color($custom-button-color);
+$custom-button-custom-corner-radius: 8px;
+$custom-button-custom-stroke-width: 4px;
+
+.custom-button--ink-color {
+  @include mdc-button-ink-color($custom-button-color);
+}
+
+.custom-button--fill-color {
```
[the remark this answers is not in the document]
Comment (acdvorak) [reply test/screenshot/mdc-button/button.test.scss]: Done

File: test/screenshot/mdc-button/custom/fill-color.test.html [hunk 17]
```diff
@@ -0,0 +1,42 @@
+<!DOCTYPE html>
```
[the remark this answers is not in the document]
Comment (acdvorak) [reply test/screenshot/mdc-button/custom/fill-color.test.html]: Done

File: test/screenshot/mdc-button/custom/stroke-width.test.html [hunk 18]
```diff
@@ -0,0 +1,42 @@
...
+
+  <body class="mdc-typography">
+    <main class="test-main">
+      <p class="test-row">
+        <span class="test-cell">
+          <button class="custom-button custom-button--stroke-width mdc-button mdc-button--stroked">Button</button>
+        </span>
+        <span class="test-cell">
+          <a class="custom-button custom-button--stroke-width mdc-button mdc-button--stroked" href="#">Link</a>
+        </span>
```
[the remark this answers is not in the document]
Comment (acdvorak) [reply test/screenshot/mdc-button/custom/stroke-width.test.html]: Done

File: test/screenshot/.stylelintrc.yaml [hunk 19]
```diff
@@ -0,0 +1,19 @@
...
+  custom-media-pattern: ^(test|custom)-.+
+  custom-property-pattern: ^(test|custom)-.+
+  selector-class-pattern:
+    - ^(test|custom)-.+
+    - resolveNestedSelectors: true
+  selector-id-pattern: ^(test|custom)-.+
+  # SCSS naming patterns, just like our CSS conventions above.
+  scss/dollar-variable-pattern:
+    - ^(test|custom)-.+
+    -
```
[the remark this answers is not in the document]
Comment (acdvorak) [reply test/screenshot/.stylelintrc.yaml]: Good catch! I copy/pasted this from our root `.stylelintrc.yaml` file. Fixed!

**Comment by acdvorak** [comment]
Added test cases for:
* `mdc-button-icon-color($color)`
* `mdc-button-horizontal-padding($padding)`
* disabled buttons (generally covering the `:not(:disabled)` logic)

**Comment by acdvorak** [comment]
- Added test cases for SVG icons
- Removed all new infrastructure

File: test/screenshot/webpack.config.js [hunk 20]
```diff
@@ -0,0 +1,58 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+const CssBundleFactory = require('../../scripts/webpack/css-bundle-factory');
```
[the remark this answers is not in the document]
Comment (acdvorak) [reply test/screenshot/webpack.config.js]: JS bundles are coming soon with whatever JS component we end up needing it for
